fastapi_login/secrets.py

```diff
@@ -0,0 +1,100 @@
+from typing import Annotated, Literal, Optional, Union
+
+from pydantic import BaseModel, Field, SecretBytes, validator
+
+try:
+    from cryptography.hazmat.backends import default_backend
+    from cryptography.hazmat.primitives import serialization
+except ImportError:  # pragma: no cover
+    _has_cryptography = False
+else:
+    _has_cryptography = True
+
+
+class RawPrivateSecret(BaseModel):
+    private_key: SecretBytes
+    password: Optional[bytes] = None
```

Comment:
Can you give an example of how this class should be used?
I assume this is one way the input can be given to the secret parameter for `LoginManager` when using asymmetric keys? If so I think it is a good idea to add a test case that uses this class in addition to the dictionary approach.

Reply:
Test cases about dictionary-secret already exists, see 

https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/tests/test_secret.py#L22-L36

and

https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/tests/test_secret.py#L61-L66


---

Each item of above is one test case, thank to pytest.

Read more at https://docs.pytest.org/en/6.2.x/parametrize.html

Reply:
Parsing are handled by pydantic model `AsymmetricSecret`, so `RawPrivateSecret` will be take care and should not use directly.

Reply:
But `RawPrivateSecret` is never invoked directly? In [`AsymmetricSecret`](https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/fastapi_login/secrets.py#L43)  data is of type `AsymmetricPairKey`, this is also mismatch of the type when passing to `AsymmetricSecretIn` as the data parameter here expects either `SecretBytes` or `RawPrivateSecret`, therefore I don't see how or why pydantic should cast the given value to either one of them.

Reply:
Because the flag [`pre=True`](https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/fastapi_login/secrets.py#L45) indicated that it is a [pre-validator](https://pydantic-docs.helpmanual.io/usage/validators/#pre-and-per-item-validators). 
Therefore `AsymmetricSecret` haven't paring raw data into `AsymmetricPairKey` yet, not until the end of pre-validator [`secret_must_be_asymmetric_private_key`](https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/fastapi_login/secrets.py#L46), `AsymmetricSecretIn` is actually dealing with raw data.

---

A step-by-step debug process will show you how data flows, it is better than plain text.

Reply:
[`secret_must_be_asymmetric_private_key`](https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/fastapi_login/secrets.py#L45-L67) is handling three different jobs:
1. validate the syntax of raw data
    - [L48](https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/fastapi_login/secrets.py#L48)
2. validate if given data is a valid asymmetric key
    - [L49-L53](https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/fastapi_login/secrets.py#L49-L53)
3. derive it's public key and cast into `AsymmetricPairKey`
    - [L55-L67](https://github.com/filwaline/fastapi_login/blob/0f0dc21d9b07f35a6b5c909c4e44d89602ecfe5d/fastapi_login/secrets.py#L55-L67)

Reply:
I think I got it now.
If one passes `{"private_key": "...", "password": "..."}` to `AsymmetricSecret`, this is then passed onto `AsymmetricSecretIn`, which in turn is recognized by pydantic as having the same fields as `RawPrivateSecret` and is thus validated as such. If all of this passes the public key is derived from the private key and then cast together as `AsymmetricPairKey`.

Reply:
> I think I got it now. If one passes `{"private_key": "...", "password": "..."}` to `AsymmetricSecret`, this is then passed onto `AsymmetricSecretIn`, which in turn is recognized by pydantic as having the same fields as `RawPrivateSecret` and is thus validated as such. If all of this passes the public key is derived from the private key and then cast together as `AsymmetricPairKey`.

Yes, that what I meant!

Reply:
Current code only accept PEM format key, do you think we should accept other formats? 
For example, DER format.

Reply:
I think PEM is sufficient for now, if the other formats are needed it shouldn't be too much trouble adding them later on.